### ISSUE TITLE
Build failure: error: redefinition of typedef ‘CK_INTERFACE’

### ISSUE THREAD

**Issue by minfrin** [OpenSC/OpenSC#2243]
### Problem Description

The master branch is currently failing to build as follows:

```
make  all-recursive
make[1]: Entering directory `/home/gatekeeper/minfrin/src/opensc/opensc-0.21.0'
Making all in etc
make[2]: Entering directory `/home/gatekeeper/minfrin/src/opensc/opensc-0.21.0/etc'
make[2]: Leaving directory `/home/gatekeeper/minfrin/src/opensc/opensc-0.21.0/etc'
Making all in src
make[2]: Entering directory `/home/gatekeeper/minfrin/src/opensc/opensc-0.21.0/src'
Making all in common
make[3]: Entering directory `/home/gatekeeper/minfrin/src/opensc/opensc-0.21.0/src/common'
  CC       compat_getopt_main.o
  CC       compat_dummy.lo
  CC       compat_strlcat.lo
  CC       compat_strlcpy.lo
  CC       compat_strnlen.lo
  CC       compat_getpass.lo
  CC       compat_getopt.lo
  CC       compat_report_rangecheckfailure.lo
  CC       compat___iob_func.lo
  CC       simclist.lo
  CCLD     libcompat.la
  CCLD     compat_getopt_main
  CC       libpkcs11.lo
In file included from libpkcs11.c:30:
../../src/pkcs11/pkcs11.h:1667: error: redefinition of typedef ‘CK_INTERFACE’
../../src/pkcs11/pkcs11.h:883: note: previous declaration of ‘CK_INTERFACE’ was here
make[3]: *** [libpkcs11.lo] Error 1
make[3]: Leaving directory `/home/gatekeeper/minfrin/src/opensc/opensc-0.21.0/src/common'
make[2]: *** [all-recursive] Error 1
make[2]: Leaving directory `/home/gatekeeper/minfrin/src/opensc/opensc-0.21.0/src'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/home/gatekeeper/minfrin/src/opensc/opensc-0.21.0'
make: *** [all] Error 2

```

Looking at the two definitions they are indeed different:

https://github.com/OpenSC/OpenSC/blame/master/src/pkcs11/pkcs11.h#L879
https://github.com/OpenSC/OpenSC/blame/master/src/pkcs11/pkcs11.h#L1667

The different definitions work because of this #define, which is conditional on CRYPTOKI_COMPAT being defined:

https://github.com/OpenSC/OpenSC/blame/master/src/pkcs11/pkcs11.h#L169

The change was done here 7 months ago:

https://github.com/OpenSC/OpenSC/commit/47151e933594b39cc790c0f4841136b83a14b034

### Proposed Resolution

Various possible fixes, one is to remove the difference between definitions, the second is to move the #define to a non conditional location.

Does anyone know what the correct fix is?

**Comment by frankmorgner** [comment]
@Jakuje ?

**Comment by Jakuje** [comment]
Sounds like bogus typedef. Can you check with the following patch:
```
diff --git a/src/pkcs11/pkcs11.h b/src/pkcs11/pkcs11.h
index f57d5f84..1b3abdaf 100644
--- a/src/pkcs11/pkcs11.h
+++ b/src/pkcs11/pkcs11.h
@@ -876,11 +876,11 @@ typedef unsigned long ck_rv_t;
 typedef ck_rv_t (*ck_notify_t) (ck_session_handle_t session,
 				ck_notification_t event, void *application);
 
-typedef struct ck_interface {
+struct ck_interface {
   char * pInterfaceName;
   void * pFunctionList;
   ck_flags_t flags;
-} ck_interface;
+};
 
 #define CKF_INTERFACE_FORK_SAFE	(0x00000001UL)
 
@@ -1169,12 +1169,12 @@ _CK_DECLARE_FUNCTION (C_GetFunctionStatus, (ck_session_handle_t session));
 _CK_DECLARE_FUNCTION (C_CancelFunction, (ck_session_handle_t session));
 
 _CK_DECLARE_FUNCTION (C_GetInterfaceList,
-		      (ck_interface *interfaces_list,
+		      (struct ck_interface *interfaces_list,
 		       unsigned long *count));
 _CK_DECLARE_FUNCTION (C_GetInterface,
 		      (unsigned char *interface_name,
 		       struct ck_version *version,
-		       ck_interface **interface,
+		       struct ck_interface **interface,
 		       ck_flags_t flags));
 
 _CK_DECLARE_FUNCTION (C_LoginUser,
```
What compiler flags do you use. We did not hit that in our ci ...

**Comment by Jakuje** [comment]
Fixed with 6063210. Closing.
